pulsar-io/elastic-search/src/main/java/org/apache/pulsar/io/elasticsearch/ElasticSearchSink.java

```diff
@@ -18,143 +18,273 @@
+                valueSchema = keyValueSchema.getValueSchema();
+                KeyValue<GenericObject, GenericObject> keyValue = (KeyValue<GenericObject, GenericObject>) record.getValue().getNativeObject();
+                key = keyValue.getKey();
+                value = keyValue.getValue();
+            } else {
+                key = record.getKey().orElse(null);
+                valueSchema = record.getSchema();
+                value = record.getValue();
+            }
+
+            String id = null;
+            if (elasticSearchConfig.isKeyIgnore() == false && key != null && keySchema != null) {
+                id = stringifyKey(keySchema, key);
+            }
 
-        if (!exists) {
-            CreateIndexRequest cireq = new CreateIndexRequest(elasticSearchConfig.getIndexName());
+            String doc = null;
+            if (value != null) {
+                if (valueSchema != null) {
+                    doc = stringifyValue(valueSchema, value);
+                } else {
+                    if (value.getNativeObject() instanceof byte[]) {
+                        // for BWC with the ES-Sink
+                        doc = new String((byte[]) value.getNativeObject(), StandardCharsets.UTF_8);
+                    } else {
+                        doc = value.getNativeObject().toString();
+                    }
+                }
+            }
 
-            cireq.settings(Settings.builder()
-               .put("index.number_of_shards", elasticSearchConfig.getIndexNumberOfShards())
-               .put("index.number_of_replicas", elasticSearchConfig.getIndexNumberOfReplicas()));
+            if (doc != null && primaryFields != null) {
+                try {
+                    // extract the PK from the JSON document
+                    JsonNode jsonNode = objectMapper.readTree(doc);
+                    id = stringifyKey(jsonNode, primaryFields);
+                } catch (JsonProcessingException e) {
+                    log.error("Failed to read JSON", e);
+                    throw e;
+                }
+            }
 
-            CreateIndexResponse ciresp = getClient().indices().create(cireq, RequestOptions.DEFAULT);
-            if (!ciresp.isAcknowledged() || !ciresp.isShardsAcknowledged()) {
-                throw new RuntimeException("Unable to create index.");
+            if (log.isDebugEnabled()) {
+                SchemaType schemaType = null;
+                if (record.getSchema() != null && record.getSchema().getSchemaInfo() != null) {
+                    schemaType = record.getSchema().getSchemaInfo().getType();
+                }
+                log.debug("recordType={} schemaType={} id={} doc={}",
+                        record.getClass().getName(),
+                        schemaType,
+                        id,
+                        doc);
             }
+            return Pair.of(id, doc);
+    } else {
+        return Pair.of(null, new String(
```

Comment:
should the record key be the `_id` here?

Reply:
we must keep the `_id` null here, because with `schemaEnable = false` the Sink behaves exactly the same way as in Pulsar 2.8.0, that is without handling the primary key of the ES document.